Notification for Auto Downloads

shortspider opened the pull request:
Resolves #2153.

shortspider commented:
The last commit (9d5e27c76aeabe79b17f7634c8f0bf46d64a0527) is purely stylistic and can be reverted if not to the project's taste.

ByteHamster commented:
Thanks! I have not looked at it too closely, yet I already have some

shortspider commented:
@ByteHamster fixed the email in the commit messages and changed generatedBySystem to initiatedByUser (agree this is better).

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/activity/OnlineFeedViewActivity.java:
```diff
@@ -225,7 +225,7 @@ public boolean onOptionsItemSelected(MenuItem item) {
     }
 
-    private void startFeedDownload(String url, String username, String password) {
+    private void startFeedDownload(String url, String username, String password, boolean initiatedByUser) {
```
This parameter seems to be always true. You can just remove it :)

shortspider replied:
Fixed.

ByteHamster commented on app/src/main/res/xml/preferences_network.xml:
```diff
@@ -35,6 +35,12 @@
+            android:enabled="true"
+            android:key="prefShowAutoDownloadReport"
+            android:summary="@string/pref_showAutoDownloadReport_sum"
+            android:title="@string/pref_showAutoDownloadReport_title"/>
```
Please change the indentation, so it matches other items in the file.

shortspider replied:
Changed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/storage/PodDBAdapter.java:
```diff
@@ -110,6 +110,7 @@
     public static final String KEY_INCLUDE_FILTER = "include_filter";
     public static final String KEY_EXCLUDE_FILTER = "exclude_filter";
     public static final String KEY_FEED_PLAYBACK_SPEED = "feed_playback_speed";
+    public static final String KEY_INITIATED_BY_USER = "initiated_by_user";
```
I do not think this is needed

shortspider replied:
It's used in `DownloadStatus` to create a staus from a cursor.

ByteHamster replied:
The cursor is used for things that come from the database. You did not update the database to contain this column. I think when creating a status from cursor, the initiatedByUser field does not matter (it is just used for the Downloads»Logs page). So you can just set it to false and remove the key here.

shortspider replied:
OK that makes sense, changed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadServiceNotification.java:
```diff
@@ -108,9 +108,11 @@ public void updateReport(List<DownloadStatus> reportQueue) {
             if (status.isSuccessful()) {
                 successfulDownloads++;
             } else if (!status.isCancelled()) {
-                createReport = true;
```
Please keep this line

shortspider replied:
Changed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadServiceNotification.java:
```diff
@@ -108,9 +108,11 @@ public void updateReport(List<DownloadStatus> reportQueue) {
-                createReport = true;
                 failedDownloads++;
             }
+            if (failedDownloads > 0 || showAutoDownloadReport && !status.isInitiatedByUser() && status.getFeedfileType() == FeedMedia.FEEDFILETYPE_FEEDMEDIA) {
```
This is only about auto download. Please do not add something related to failed downloads here.

shortspider replied:
Changed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadStatus.java:
```diff
@@ -41,62 +41,49 @@
 	 * FEEDFILETYPE_FEEDIMAGE or FEEDFILETYPE_FEEDMEDIA
 	 */
     private final int feedfileType;
+    private final boolean initiatedByUser;
```
Instead of adding the field to DownloadStatus, please use `status.getRequest().isInitiatedByUser()`. Then we do not need another copy of the variable.

shortspider replied:
There is no `getRequest` method that I can see. I can add it but not all status' are created with a request, and even so that seems like overkill. Or were you just talking about the constructor that takes a request? I'll clean that up either way.

ByteHamster replied:
Oh you are right. Then it's OK this way.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/storage/DBTasks.java:
```diff
@@ -199,11 +199,12 @@ private static void refreshFeeds(final Context context,
+                                               DownloadError.ERROR_REQUEST_ERROR,
+                                               false,
+                                               e.getMessage(),
+                                               false)
```
Users can also press the refresh button. I think this method might be called when users press the refresh button, so it is not always initiated by the system. Not sure what to pick here, though.

shortspider replied:
Good catch, managed to get it sorted out.

shortspider commented:
@ByteHamster the two failing tests pass locally for me. I there a way I can get checks to run again or do I need to make a dummy commit?

ByteHamster commented:
> the two failing tests pass locally for me. I there a way I can get checks to run again or do I need to make a dummy commit?

Unfortunately, GitHub Actions does not allow to re-run tests if you do not have write permissions to the whole repo. Just ping me if it fails again after adding the review changes

shortspider commented:
@ByteHamster OK changes made and all tests passing this time.

ByteHamster commented:
I just noticed that the notification will be displayed in the "error" notification channel. That could confuse users. I think we should add another notification channel for the auto download report.

shortspider commented:
@ByteHamster I added the channel and changed one existing string to make it applicable for error reports and auto download reports.

One thing occured to me as well was what should happen when the notification for auto downloads is clicked. Currently it will take you to the downloads page, but I think it would make more sense to take you to the queue page. What do you think? If you agree should I modify the `getReportNotificationContentIntent` method to handle both cases or create a new method entirely?

ByteHamster commented:
I am just wondering if #213 (Notify new episodes) would basically serve the same purpose as this PR but could help more users. That way, we would only need 2 notification settings instead of 3. I am trying to keep the number of additional settings as low as possible because more settings make the app harder to use for new users.

shortspider commented:
I agree that keeping the number of settings down makes sense. Should there be one setting but two notifications, one to indicate there is a new episode and another to show that it's been downloaded? If the user is setup to only download over wifi then the indication of a new episode doesn't mean that it's been downloaded.

The other question I have is where should the notificaion take you when clicked? I would say the queue if we are going to have one notification for only new episodes. Going to the queue will tell the user if the episode has been downloaded or not.

In terms of implementation, it looks like we would throw away almost all of what I have added and instead add some logic to `DownloadServiceCallbacksImpl.onFeedParsed` to create the notification when there are new feed items. Does that make sense to you? If so, do you mind if I leave some of the formatting changes I've added, even if the new properties and whatnot are removed?

ByteHamster commented:
> If the user is setup to only download over wifi then the indication of a new episode doesn't mean that it's been downloaded.

Hmm that's right.

> In terms of implementation, it looks like we would throw away almost all of what I have added

My suggestion was just an idea. You are right, it would be bad to throw away everything. Maybe we can keep both settings but change the settings menu, so we don't clutter the interface. It could be a list with checkboxes like the one on Settings/Network/Mobile updates. Alternatively, we could add a new settings screen for notifications.

shortspider commented:
I am OK to toss away what I've done if it means a better app in the end. I just need some guidance from you in what you think would be best.

I currently have two questions:

1. Regardless of what we notify on (download or new episode available) should pressing the notification take you to downloads, the queue, the podcast etc?
2. Do we want to notify on episodes downloading or episodes being available or both?

I honestly don't really have an opinion on 2, I think users who are setup not to download on mobile will be OK if they are notified when episodes are available but I just wanted to point that out.

With regards to 1, IMO pressing the notification should go to the queue since that makes the most sense when multiple episodes are downloaded (say overnight). If you are not on wifi and won't download over mobile then the queue won't have the episodes and that's fine (again IMO).

shortspider commented:
@ByteHamster any thoughts?

ByteHamster commented on core/src/main/res/values/strings.xml:
```diff
@@ -245,7 +245,7 @@
     <string name="download_canceled_msg">Download canceled</string>
     <string name="download_canceled_autodownload_enabled_msg">Download canceled\nDisabled <i>Auto Download</i> for this item</string>
-    <string name="download_report_title">Downloads completed with error(s)</string>
+    <string name="download_report_title">Downloads completed</string>
```
I think there should be two strings. One for auto downloads completed and one for failed downloads.

shortspider replied:
Changed.

ByteHamster commented on core/src/main/res/values/strings.xml:
```diff
@@ -781,6 +783,8 @@
     <string name="notification_channel_error">Errors</string>
     <string name="notification_channel_error_description">Shown if something went wrong, for example if download or gpodder sync fails.</string>
+    <string name="notification_channel_auto_download">Auto Downloads</string>
+    <string name="notification_channel_error_auto_download">Shown when episodes are automatically downloaded.</string>
```
Maybe "have been" downloaded? Not a native English speaker but the notification is shown after the downloads were completed. "Are" sounds like the download is still running.

shortspider replied:
Changed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/storage/DownloadRequester.java:
```diff
@@ -203,25 +203,26 @@ public synchronized void downloadFeed(Context context, Feed feed, boolean loadAl
 
     @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
-    public synchronized void downloadMedia(boolean performAutoCleanup, @NonNull Context context,
+    public synchronized void downloadMedia(boolean performAutoCleanup, @NonNull Context context, boolean initiatedByUser,
```
Please break to a new line. The maximum line length should be 120 characters. This is currently not enforced by the CI because other parts of the app do not comply but I want new code to be consistent.

shortspider replied:
Changed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadRequest.java:
```diff
@@ -107,6 +134,7 @@ public void writeToParcel(Parcel dest, int flags) {
         dest.writeString(nonNullString(password));
         dest.writeByte((mediaEnqueued) ? (byte) 1 : 0);
         dest.writeBundle(arguments);
+        dest.writeByte(initiatedByUser ? (byte)1 : 0);
```
Please add a space after `(byte)`

shortspider replied:
Changed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadRequest.java:
```diff
@@ -38,50 +39,76 @@ public DownloadRequest(@NonNull String destination,
+             password,
+             false,
+             arguments,
+             initiatedByUser);
```
To be honest, I think this is harder to read than before. This nearly fills the screen of my laptop vertically and is basically empty horizontally. Please break the lines, so that one line has more than one parameter.

shortspider replied:
Changed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadRequest.java:
```diff
@@ -258,6 +287,8 @@ public boolean isMediaEnqueued() {
         return mediaEnqueued;
     }
 
+    public boolean isInitiatedByUser() { return initiatedByUser; }
```
Please add line breaks

shortspider replied:
Changed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadStatus.java:
```diff
@@ -109,18 +95,41 @@ public static DownloadStatus fromCursor(Cursor cursor) {
+								  DownloadError.fromCode(cursor.getInt(indexReason)),
+								  new Date(cursor.getLong(indexCompletionDate)),
+								  cursor.getString(indexReasonDetailed),
+								  false);
```
This file seems to be indented with tabs instead of spaces. While you are already touching big parts of the file, it would make sense to let Android Studio convert the file to spaces instead of tabs. That would also fix the strange indention above.

shortspider replied:
Changed.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadService.java:
```diff
@@ -205,9 +205,10 @@ public void onDestroy() {
-                && UserPreferences.showDownloadReport()) {
-            notificationManager.updateReport(reportQueue);
+                && (UserPreferences.showDownloadReport() || showAutoDownloadReport)) {
+            notificationManager.updateReport(reportQueue, showAutoDownloadReport);
```
Just wondering... The two kinds of notifications are now quite different (Channel, text, maybe even icon). Would it make sense to extract the auto download notification to a new class instead of handling everything in the existing `notificationManager`? I have not checked if this will lead to a lot of code duplication, though.

shortspider replied:
I'm not 100% sure what you mean here. `DownloadService` seems to be the right place to kick off a notification and `notificationManager` (to me anyway) seems like a class that will generate whatever notification is required for a given set of parameters. IMO if we did create another class there would be a fair bit of duplication.

ByteHamster commented on core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadServiceNotification.java:
```diff
@@ -107,29 +107,28 @@ public void updateReport(List<DownloadStatus> reportQueue) {
+
+            NotificationCompat.Builder builder = new NotificationCompat.Builder(context, channelId);
+            builder.setTicker(context.getString(titleId))
+                   .setContentTitle(context.getString(R.string.download_report_content_title))
```
Please use the same string for ticker and content title. Also, I think it would be easier to read the code if you copied the notification builder code: `if(autoDownloadReport){...}else{...}` instead of using the `?` operator in basically every line.

shortspider replied:
Agreed, done.

shortspider commented on core/src/main/java/de/danoeh/antennapod/core/service/download/handler/FeedParserTask.java:
```diff
@@ -78,12 +78,12 @@ public FeedHandlerResult call() {
         } else {
-            downloadStatus = new DownloadStatus(feed, request.getTitle(),
-                    reason, successful, reasonDetailed);
+            downloadStatus = new DownloadStatus(feed, feed.getHumanReadableIdentifier(), reason, successful,
```
Changed.

shortspider commented on core/src/main/res/values/strings.xml:
```diff
@@ -782,6 +785,8 @@
     <string name="notification_channel_error">Errors</string>
     <string name="notification_channel_error_description">Shown if something went wrong, for example if download or gpodder sync fails.</string>
+    <string name="notification_channel_auto_download">Auto Downloads</string>
+    <string name="notification_channel_error_auto_download">Shown when episodes have been automatically downloaded.</string>
```
Changed.

shortspider commented on core/src/main/res/values/strings.xml:
```diff
@@ -477,6 +478,8 @@
     <string name="pref_showDownloadReport_title">Show Download Report</string>
     <string name="pref_showDownloadReport_sum">If downloads fail, generate a report that shows the details of the failure.</string>
+    <string name="pref_showAutoDownloadReport_title">Show Auto Download Report</string>
+    <string name="pref_showAutoDownloadReport_sum">Generate a report that shows the details of auto downloaded episodes.</string>
```
Agreed, changed.

shortspider commented on core/src/main/res/values/strings.xml:
```diff
@@ -245,6 +245,7 @@
     <string name="download_canceled_msg">Download canceled</string>
     <string name="download_canceled_autodownload_enabled_msg">Download canceled\nDisabled <i>Auto Download</i> for this item</string>
     <string name="download_report_title">Downloads completed with error(s)</string>
+    <string name="auto_download_report_title">Downloads completed</string>
```
Changed.

shortspider commented on core/src/main/java/de/danoeh/antennapod/core/util/gui/NotificationUtils.java:
```diff
@@ -62,4 +64,14 @@ private static NotificationChannel createChannelError(Context c) {
+                                            CHANNEL_ID_AUTO_DOWNLOAD,
+                                            c.getString(R.string.notification_channel_auto_download),
+                                            NotificationManager.IMPORTANCE_DEFAULT);
+        mChannel.setDescription(c.getString(R.string.notification_channel_downloading_description));
```
Done.

shortspider commented on core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadServiceNotification.java:
```diff
@@ -107,29 +107,28 @@ public void updateReport(List<DownloadStatus> reportQueue) {
         for (DownloadStatus status : reportQueue) {
             if (status.isSuccessful()) {
                 successfulDownloads++;
+                createReport = createReport || showAutoDownloadReport && !status.isInitiatedByUser() && status.getFeedfileType() == FeedMedia.FEEDFILETYPE_FEEDMEDIA;
```
Agreed.

shortspider commented on app/src/main/java/de/danoeh/antennapod/config/DownloadServiceCallbacksImpl.java:
```diff
@@ -38,13 +39,15 @@ public PendingIntent getAuthentificationNotificationContentIntent(Context contex
 
     @Override
-    public PendingIntent getReportNotificationContentIntent(Context context) {
+    public PendingIntent getReportNotificationContentIntent(Context context, boolean autoDownloadReport) {
```
Done.

shortspider commented:
Hey no problem on the response speed, it's a very intersting time right now lol. Hope you're safe and healthy.

Do you want me to resolve these comments as I go? At my work we leave it up to the reviewer to resolve that the comment has been addressed to their statification but not sure if that is how this project does things.

ByteHamster commented:
> Hey no problem on the response speed, it's a very intersting time right now lol. Hope you're safe and healthy.

I am fine, thanks :) Hope that you are, too.

> Do you want me to resolve these comments as I go? At my work we leave it up to the reviewer to resolve that the comment has been addressed to their statification but not sure if that is how this project does things.

I just marked them as resolved.

---

I am finally happy with the code structure :D Just noticed three little things when testing on my device:

- You currently use the `stat_notify_sync` icon for the notification. Would be cool to have a `stat_notify_sync_success` (just like `stat_notify_sync_error`) with a check mark. Otherwise, users can not tell the two notifications apart on a quick look. Bonus points if you create a vector image instead of the current png files ;)
- What do you think about displaying the list of downloaded episodes in the notification? Similar to the notification that is displayed while downloading:
  <img width="250" src="https://user-images.githubusercontent.com/5811634/77260443-5e336500-6c88-11ea-99a7-57ead2d33fad.png" />
- You currently use the same notification id for both the error report and the success dialog. This means they replace each other and only one of them can be displayed at a time. I think it would make sense to use a unique id.

keunes commented:
> > Regardless of what we notify on (download or new episode available) should pressing the notification take you to downloads, the queue, the podcast etc?
> 
> I think the queue would be a good location.

When a new episode is available, bringing users to the queue might not make much sense? I feel in this case the podcast would make more sense (from there they can download, mark as favourite, in future maybe add it to a specific playlist/category/folder...). For downloads the queue seems most appropriate indeed.

ByteHamster commented:
> When a new episode is available, bringing users to the queue might not make much sense?

This PR is about auto downloaded episodes, not new episodes. By default, all downloaded episodes get added to the queue, so the notification will show the episodes after clicking.

shortspider commented:
@ByteHamster so I've updated the notification id, the text content and the icon. The icon is from [here](https://material.io/resources/icons/?icon=check_circle_outline&style=outline) which I think is OK to use, seems to be Apache 2.0 licence.

ByteHamster commented:
Thanks. I will change the icon to have a circular arrow like the refresh icon after merging. I have to fiddle with icons anyway to convert the other icons to vector format.

ByteHamster commented:
:tada: Merged :) Thanks a lot for your pull request!

shortspider commented:
Awesome thank you! Good catch on that StringBuilder!
